cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java

```diff
@@ -849,7 +852,155 @@ public Object run() throws Exception
         });
     }
 
-////    @Test
+
+
+    @Test
+    public void testGetMultipleUser() throws Exception
+    {
+        // add user using X500Principal
+        // was MultiAccount*, Multi*
+        final String username = "DuplicateCadcUser2";
+        log.info(username);
+
+        try {
+            // This allows multiple users with same DN to be created. This test requires
+            // that scenario, in order to test that getUser works correctly under what
+            // is generally an abnormal condition.
+
+            // This test is to test a hack in the system to prevent issues where multiples
+            // of users are created. (s2255)
+            System.setProperty(LdapUserDAO.SUPPRESS_CHECKUSER_KEY, "true");
+
+            final X500Principal userID = new X500Principal("cn=" + username + ",ou=cadc,o=hia,c=ca");
+            final HttpPrincipal testHttpPrincipal = new HttpPrincipal(username);
+
+            log.info("user DN: " + config.getUsersDN());
+            final DNPrincipal dnPrincipal = new DNPrincipal("uid=" + username + "," + config.getUsersDN());
+            Subject subject = new Subject();
+            subject.getPrincipals().add(dnPrincipal);
+
+            Subject.doAs(subject, new PrivilegedExceptionAction<Object>() {
+                public Object run() throws Exception {
+                    final LdapUserDAO userDAO = getUserDAO();
+                    User testUser = new User();
+
+                    // Test to see if user already exsits.
+                    try {
+                        testUser = userDAO.getUser(userID);
+                    } catch (UserNotFoundException unfe) {
+                        testUser.getIdentities().add(userID);
+                        testUser.getIdentities().add(testHttpPrincipal);
+                        testUser.personalDetails = new PersonalDetails("Multi", "MultiAccountCadcUser");
+                        testUser.personalDetails.email = username + "@canada.ca";
+                        log.debug("Test user " + username + " not found, creating...");
+                        UserRequest userRequest = new UserRequest(testUser, "password".toCharArray());
+                        userDAO.addUserRequest(userRequest);
+                        userDAO.approveUserRequest(testHttpPrincipal);
+                    }
+
+                    // Check to see if there are multiple users already existing.
+                    List<User> userList = userDAO.getAllUsers(userID, config.getUsersDN());
+
+                    if (userList.size() < 6) {
+                        log.debug("Creating multiple users (5 more)");
+
+                        // Run this 5x to create multiple users
+                        for (int i = 0; i < 5; i++) {
+                            try {
+                                userDAO.addUser(testUser);
```

Comment:
It appears to me that the testUser are all identical (with or without HttpPrincipal) which is not exactly the main scenario of one real user with multiple duplicates that don't have HttpPrincipal

Reply:
The key thing is userDAO.addUser is the internal method (inserts user directly into final tree, with $EXTERNAL-CN as LDAP-USER-NAME. the 'addUserRequest/approveUserRequest' cycle inserts the user information correctly into the ldap entry. HttpPrincipal existing or not isn't part of the check.

Reply:
OK, I see. But I would say it's a bit risky to have the method you are testing (getUser) part of the setup of the test (line 889).